db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/subquery/aggregate/AggregateSubqueryPredicate.java

```diff
@@ -88,6 +118,19 @@ private boolean doWeHandle(SubqueryNode subqueryNode) throws StandardException {
                 (subquerySelectNode.getGroupByList() == null || subquerySelectNode.getGroupByList().isEmpty()))
             return false;
 
+        /*
+        * the aggregate must not be value-producing, flattening such aggregates could lead to wrong results since the resulting inner join
+        * could miss some rows.
+        * For example:
+        *    select a,b from table1 where 0 = (select count(table2.a) from table2 where table2.a = table1.a)
+        *     if table2 is empty, then count() will return 0 (value-producing), causing the where condition to be satisfied.
+        *     however, if we flatten, then we end up with something like this:
+        *    select table1.a,b from table1, (select table2.a, count(table2.a) from table2 group by table2.a having count(table2.a) = 0) X where table1.a = X.a
+        *     which will return empty result since the RHS of the join is empty, and the join type is inner.
+        */
+        if(isValueProducingExpr(resultColumns.elementAt(0))) {
+            return false;
+        }
```

Comment:
I would have a loop checking all `resultColumns` instead of just check the first one.

Yes, we don't support tuple comparison (e.g., `(0, 1) = (select sum(a), avg(a) from ...)`) yet. But given that we already moved one step further to support `(a, b,) in (select x, y from ...)`, tuple comparison is not far anymore. It doesn't hurt to anticipate that.

Reply:
Good point, I just followed the current assumptions BTW (lines 108-111):
```
        ResultColumnList resultColumns = subqueryResultSet.getResultColumns();
        if (resultColumns.size() != 1) {
            return false;
        }
```